File: FangZhiHu/CommonLib/src/main/java/network/BaseSubscriber.java
package network;


import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;

import io.reactivex.subscribers.ResourceSubscriber;
import network.exception.ErrorCode;
import network.exception.SendException;
import qj.ccmtjf.com.lib.api.ApiResponse;


/**
 * 网络请求返回数据分发处理
 * Created by yuantongqin on 2017/8/7.
 * modify by liyanxi on 2017/9/27.{@link }新增后端data数据传递子类
 */

public abstract class BaseSubscriber<T> extends ResourceSubscriber<T> {
    private static final String TAG = BaseSubscriber.class.getSimpleName();
    private  Gson gson;

    public BaseSubscriber() {
        super();
        gson = new GsonBuilder().create();
    }

    @Override
    public void onStart() {
        super.onStart();
        //检查是否有网络的信息

    }

    @Override
    public void onComplete() {
        onFinish();
    }

    public abstract void onSuccess(T t);

    public abstract void onFailure(ApiResponse<String> error);

    public void onFinish() {
    }

    @Override
    public void onNext(T t) {
        onSuccess(t);
    }

    @Override
    public void onError(Throwable e) {
        //同意处理错误信息
        try{
            if (e instanceof SendException) {
                SendException exce = (SendException) e;
                Log.e(TAG, exce.getCode() + "====" + exce.getData());
                if(exce.customState == ErrorCode.ERROR_STATE_ZERO){
                    if(gson == null){
                        gson = new GsonBuilder().create();
                    }
                    Type type = new TypeToken<ApiResponse<Object>>() {
                    }.getType();
                    ApiResponse<Object> error = gson.fromJson(exce.getData(), type);
                    String da = "";
                    if(error.getData() != null){
                        da = error.getData().toString();
                    }
                    onFailure(new ApiResponse<>(da, error.getMessage(), exce.getCode()));
                }else{
                    ApiResponse<String> error = new ApiResponse<>(exce.getData(), exce.getMsg(), exce.getCode());
                    onFailure(error);
                }

            } else {
                String message = "请稍后再试";
                Log.e(TAG,e.toString());
                if (e instanceof IOException) {
                    message = "网络链接异常，请稍后再试";
                }
                ApiResponse<String> error = new ApiResponse<>(ErrorCode.NETWORK_ERROR, message);
                onFailure(error);
            }
            onFinish();
        }catch (Exception e1){
            e1.printStackTrace();
        }


    }


}

File: FangZhiHu/CommonLibKotlin/src/main/java/network/BaseSubscriber.kt
package network

import android.util.Log
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import io.reactivex.observers.ResourceObserver
import io.reactivex.subscribers.ResourceSubscriber
import network.converter.GsonConverterFactory
import network.entity.ApiResponse
import network.excepition.ErrorCode
import network.excepition.SendException
import java.io.IOException

/**
 * Created by tongqinyuan on 2018/3/7.
 */
abstract class BaseSubscriber<T> : ResourceSubscriber<T>() {
    private val TAG:String = BaseSubscriber::class.java.simpleName
    private var mGosn:Gson ?= null
    init {
        mGosn = GsonBuilder().create()
    }

    override fun onStart() {
        super.onStart()
    }

    override fun onNext(t: T) {
        onSuccess(t)
    }

    override fun onComplete() {
        onFinish()
    }

    fun onFinish() {}

    override fun onError(t: Throwable?) {
        try {
            if (t is SendException) {
                val exce = t as SendException
                Log.e(TAG,  "===$exce.code=" + exce.data)
                if (exce.customState == ErrorCode.ERROR_STATE_ZERO) {
                    if (mGosn == null) {
                        mGosn = GsonBuilder().create()
                    }
                    val type = object : TypeToken<ApiResponse<Any>>() {

                    }.type
                    val error :ApiResponse<Any> = mGosn!!.fromJson(exce.data, type)
                    var da  = ""
                    if (error.data != null) {
                        da = error.data.toString()
                    }
                    onFailure(ApiResponse(da, error.message!!, exce.code))
                } else {
                    val error = ApiResponse<String>(exce.data!!, exce.msg!!, exce.code)
                    onFailure(error)
                }

            } else {
                var message = "请稍后再试"
                Log.e(TAG, t.toString())
                if (t is IOException) {
                    message = "网络链接异常，请稍后再试"
                }
                val error = ApiResponse<String>(ErrorCode.NETWORK_ERROR, message)
                onFailure(error)
            }
            onFinish()
        } catch (e1: Exception) {
            e1.printStackTrace()
        }


    }

    abstract fun onSuccess(t:T)

    abstract fun onFailure(error :ApiResponse<String>)
}